core/chains/evm/logpoller/log_poller_test.go

```diff
@@ -1416,7 +1416,7 @@ func TestTooManyLogResults(t *testing.T) {
 		from := fq.FromBlock.Uint64()
 		to := fq.ToBlock.Uint64()
 		if to-from >= 4 {
-			return []types.Log{}, &clientErr // return "too many results" error if block range spans 4 or more blocks
+			return []types.Log{}, clientErr // return "too many results" error if block range spans 4 or more blocks
```

Comment:
What happened here? Is this declared in a way that makes it easy to make that mistake?

Reply:
Actually I'm glad you asked, there might still be something wrong here. The Error field returned by geth directly in a jsonMessage is set to their own internal type of `&jsonError`:
https://github.com/smartcontractkit/go-ethereum/blob/master/rpc/json.go#L109-L123

Then our client wrapper converts it at some point into our own type of `JsonError`, but confusingly... we redefine geth's `jsonError` in 3 different places, twice as a struct and once as an interface:
```
core/chains/evm/client/errors.go:type JsonError struct {
core/chains/evm/client/simulated_backend_client.go:     // type JsonError struct {
core/scripts/chaincli/handler/report.go:type JsonError interface {
```
Based on these, it looks like returning `&clientErr` was the right thing to do after all.  But if that's true, then the check for `Error.As()` will fail unless we pass it a `**JsonError` which doesn't seem right either.  Maybe I shouldn't be checking against a concrete type anyway, it's too brittle?  I think I need to investigate this more and figure out what the different ways are in which this error might be returned.

Reply:
I finally had a chance to go back and look through geth's error handling more carefully to sort this out.  The number of different ways in which these json rpc errors can get transformed and wrapped as they are propagating up the stack is enormous. The type changes several times, sometimes involving a pointer and sometimes not.

There is an interface defined for this purpose in geth, `rpc.Error`, which is designed to test whether something has an error code attached to it, which is all we need.  The only problem with that was that the JsonError type we define that it can get transformed into under some conditions doesn't actually satisfy this. Since it's clear from the comments that the main purpose of us defining this type was to mimick the `jsonError` type in geth which isn't exported, I fixed this by adding the `ErrorCode()` method so that they both behave the same. (Also noticed a pointer receiver on the String method, which I think was not supposed to be there.)

While looking through the geth code, I noticed another error code it can return which it would make sense for us to handle in the same way, so I've added that as well. And added a regression test for this bug of logging a critical error when there's nothing critical going on.